feat: use crypto provider

Qumeric opened the pull request:
Use recently introduced 

```
use alloy_consensus::crypto::CryptoProvider
```
and
```
use revm::precompile::Crypto
```

to use openvm-optimized cryptography without custom revm fork.

Crypto implementation is taken from https://github.com/axiom-crypto/revm/pull/76/files and adapted

* [Baseline run](https://github.com/axiom-crypto/openvm-reth-benchmark/blob/gh-pages/benchmarks-dispatch/refs/heads/main/reth-5d8c6b110da65c9a67ec0069727fda0fafd6b7fe-ecae167dff2ae3e38638889140a8d22a6620cdc85093de97c2db3c3f6d573d79.md)
* [This PR run](https://github.com/axiom-crypto/openvm-reth-benchmark/blob/gh-pages/benchmarks-dispatch/refs/heads/val/update-revm-reth/reth-3e771553323d66908ea7592160ea89987557faaa-ecae167dff2ae3e38638889140a8d22a6620cdc85093de97c2db3c3f6d573d79.md)

---
Closes INT-4670

jonathanpwang commented:
Update: I think revm has released this Crypto provider feature https://github.com/bluealloy/revm/releases/tag/v84

Qumeric commented:
Yes, revm released, I use release. But reth has not released stable version supporting it, so I use main branch

Qumeric commented:
Ok so I rebased again and updated openvm

On app-prove regression is quite noticeable:
* main total_cells_used `4,491,246,156` 
* this total_cells_used `4,786,702,566` (+6.5%)

Qumeric commented:
Looking at flamegraph, it's really not clear why it is slower. 
2 Things I noticed:
1. reth apparently improved deserialization time for some things (not MPT). So total time should be 1% better or so but regression elsewhere is large
2. `MptTrie::calc_reference` takes 49% more samples. Seems suspicious?

shayanh commented:
While I was waiting for `terraform apply`, I updated this to reth 1.7.0 and resolved the conflicts in `val/update-revm-reth-1.7.0` branch. To clarify, I made a new branch because I didn't want to suddenly force push here.

Benchmarks:
- [`main`](https://github.com/axiom-crypto/openvm-reth-benchmark/actions/runs/17809406622)
- [`val/update-revm-reth-1.7.0`](https://github.com/axiom-crypto/openvm-reth-benchmark/actions/runs/17809020667)

Unfortunately there is a bit of a performance regression which I believe must be because of the changes in revm and reth.

|                           | `main`        | `val/update-revm-reth-1.7.0` | **% Δ** |
|---------------------------|---------------|------------------------------|---------|
| Proof Time (s)            | 163.17        | 166.37                       | **+1.96%** |
| `total_cells_used`        | 3,999,234,686 | 4,274,946,962                | **+6.90%** |
| `execute_e1_time_ms`      | 719           | 744                          | **+3.48%** |
| `execute_preflight_insns` | 128,251,685   | 132,294,410                  | **+3.15%** |

The regression is relatively small and I suggest we should go ahead and merge this because:
- We need to update reth for the improved client input generation.
- At some point we should update reth anyway.

Qumeric commented:
Have we compared with patched revm? I think we should at least check that crypto provider itself is not slower. I think @jonathanpwang patched revm exactly for this purpose?

Also, I think it would be great if we could pinpoint the source of the regression. Because possibly it would be fixable. Otherwise we would likely just forget and potentially miss optimization opportunity.

shayanh commented:
I don't we have an updated revm patch.

shayanh commented:
Updated reth to v1.8.3 in `val/update-revm-reth-1.8.3` branch.

jonathanpwang commented:
Likely `ruint` isn't patched anymore because of version differences, so it's a little more fair to compare with https://github.com/axiom-crypto/openvm-reth-benchmark/pull/521 (which I will now merge)

jonathanpwang commented on crates/executor/client/src/crypto.rs:
```diff
@@ -0,0 +1,264 @@
+
+/// OpenVM custom crypto implementation for faster precompiles
+#[derive(Debug, Default)]
+struct OpenVMCrypto;
```
nit: `Vm` for rust naming convention

jonathanpwang commented on Cargo.toml:
```diff
@@ -146,6 +148,20 @@ openvm-sdk = { git = "https://github.com/openvm-org/openvm.git", branch = "main"
+
+openvm-kzg = { git = "https://github.com/axiom-crypto/openvm-kzg.git", branch = "val/openvm-main", default-features = false }
+
+k256 = { git = "https://github.com/openvm-org/openvm.git", branch = "main" }
```
it's better if we name this `openvm-k256` when not patching to avoid confusion

jonathanpwang commented:
superseded by https://github.com/axiom-crypto/openvm-reth-benchmark/pull/523
